### PR TITLE
performance patch of JsonConverterHelper.cs

### PR DESCRIPTION
**Pull request by PiDiBi**
as far as we use this in azure web apps and it makes us problems I'm kindly suggesting this performance update
- removed LINQ
- regex split is compiled singeltone now

BenchmarkDotNet=v0.13.1, OS=Windows 10.0.22000
Intel Xeon W-2133 CPU 3.60GHz, 1 CPU, 12 logical and 6 physical cores
.NET SDK=6.0.101
  [Host]     : .NET 5.0.13 (5.0.1321.56516), X64 RyuJIT  [AttachedDebugger]
  DefaultJob : .NET 5.0.13 (5.0.1321.56516), X64 RyuJIT

|        Method |     Mean |    Error |   StdDev |
|-------------- |---------:|---------:|---------:|
| OriginalSplit | 985.5 ns | 19.27 ns | 47.63 ns |
|     NewSplit1 | 478.3 ns |  9.56 ns | 12.44 ns |

On Dec 11, 2021, at 10:09 AM, Feng Yuan <fyuan@microsoft.com> wrote: 
Found a super expensive regular expression stack in NetworkWatcherTrafficAnalytics-Prod: 
image
GetPropertyName is costing 4.1 million CPU samples, 14.84% of total CPU samples in the stream I downloaded.

testing code
```

using BenchmarkDotNet.Attributes;
using BenchmarkDotNet.Running;
using System;
using System.Linq;
using System.Text.RegularExpressions;

namespace Becnhmarks.ConsoleX
{
    internal class Program
    {
        static void Main(string[] args)
        {
            BenchmarkRunner.Run<MyClass>();
        }
    }

    public class MyClass
    {
        private readonly string data = @"parent1\..parent2\..parent3\.";
        private string[] parent;
        
        private static readonly Regex splitCompiledRegex = new Regex(@"(?<!\\)\.", RegexOptions.Compiled);

        [Benchmark]
        public string OriginalSplit() => GetPropertyNameOriginal(data, out parent);

        [Benchmark]
        public string NewSplit1() => GetPropertyNameNew1(data, out parent);


        public static string GetPropertyNameOriginal(string name, out string[] parentPath)
        {

            string propertyName = name;
            parentPath = new string[0];

            if (!string.IsNullOrEmpty(propertyName))
            {
                string[] hierarchy = Regex.Split(propertyName, @"(?<!\\)\.")
                    .Select(p => p?.Replace("\\.", ".")).ToArray();
                if (hierarchy.Length > 1)
                {
                    propertyName = hierarchy.Last();
                    parentPath = hierarchy.Take(hierarchy.Length - 1).ToArray();
                }
            }

            return propertyName;
        }

        public static string GetPropertyNameNew1(string name, out string[] parentPath)
        {

            string propertyName = name;
            parentPath = new string[0];

            if (!string.IsNullOrEmpty(propertyName))
            {
                string[] hierarchy = splitCompiledRegex.Split(propertyName);
                for (int i = 0; i < hierarchy.Length; i++)
                {
                    hierarchy[i] = hierarchy[i]?.Replace("\\.", ".");
                }

                if (hierarchy.Length > 1)
                {
                    propertyName = hierarchy[hierarchy.Length - 1];
                    Array.Resize(ref hierarchy, hierarchy.Length - 1);
                    parentPath = hierarchy;
                }
            }

            return propertyName;
        }

    }
}

```

### CODE REVIEW

File: sdk/mgmtcommon/ClientRuntime/ClientRuntime/Serialization/JsonConverterHelper.cs [hunk 1]
```diff
@@ -113,19 +115,24 @@ public static string GetPropertyName(this JsonProperty property, out string[] pa
 
             if (!string.IsNullOrEmpty(propertyName))
             {
-                string[] hierarchy = Regex.Split(propertyName, @"(?<!\\)\.")
-                    .Select(p => p?.Replace("\\.", ".")).ToArray();
+                string[] hierarchy = splitCompiledRegex.Split(propertyName);
```
Comment (ArthurMa1978) [inline sdk/mgmtcommon/ClientRuntime/ClientRuntime/Serialization/JsonConverterHelper.cs:118]: As this func is frequently invoked, so consider not use Regex but just a single loop to complete both split and replace.

File: sdk/mgmtcommon/ClientRuntime/ClientRuntime/Serialization/JsonConverterHelper.cs [hunk 2]
```diff
@@ -113,19 +115,24 @@ public static string GetPropertyName(this JsonProperty property, out string[] pa
...
-                    propertyName = hierarchy.Last();
-                    parentPath = hierarchy.Take(hierarchy.Length - 1).ToArray();
+                    propertyName = hierarchy[hierarchy.Length - 1];
+                    Array.Resize(ref hierarchy, hierarchy.Length - 1);
+                    parentPath = hierarchy;
                 }
             }
 
-            return propertyName;            
+            return propertyName;
```
Comment (ArthurMa1978) [inline sdk/mgmtcommon/ClientRuntime/ClientRuntime/Serialization/JsonConverterHelper.cs:132]: good catch :)

**Review by ArthurMa1978 — APPROVED** [review]
LGTM, and remove the Regex should be better

File: sdk/mgmtcommon/ClientRuntime/ClientRuntime/Serialization/JsonConverterHelper.cs (hunk 1 above)
Comment (PiDiBi) [reply sdk/mgmtcommon/ClientRuntime/ClientRuntime/Serialization/JsonConverterHelper.cs:118]: what I understand from the code the split with regex is used to split by '.' and skip splitting by '\.' as far it's escaped '.'
so not sure how can you in one loop split by '.' with skipping '\.' not using the regex and then replacing '\.' with '.'
if you first split by '.' and then you need to go thru loop and join subsequent array items ending with '\' ..
I guess this solution us much better than it was so I would let it for someone from your team to polish it if needed

**Comment by AlexGhiondea** [comment]
@PiDiBi thanks for your PR!

Do you happen to have the benchmarks for allocations? Did they change?

**Comment by PiDiBi** [comment]
should I delete this PR? nobody interested in performance improvements?

**Comment by PiDiBi** [comment]
halloooooo, anyone here interested in performance improvement?

**Pull request merged**